README.md
# bbc-weather-cli
A cli tool to fetch weather reports from the BBC's JSON API. 

I originally wrote this for a custom waybar module on my sway wm + waybar desktop. And also because I didn't want to parse that BBC XML RSS feed.

## Using with Waybar

I have an example waybar config in `waybar/example-config.json`. To use `bbc-weather-cli` with waybar, add the `custom/bbcweather` into your waybar config's modules declaration.

## Requirements

A font with adequate icon sets, such as [Nerd Fonts](https://www.nerdfonts.com/).

## Usage

`bbc-weather-cli` prints a waybar-formatted JSON to stdout by default.

``` shell
$ bbc-weather-cli
{"text":"  11","alt":" 11 kph E : 1% | Sunny intervals and light winds","tooltip":"<big>London, Greater London</big>\nNow:  11 kph E : 1% | Sunny intervals and light winds\n\nSunday, 2022-03-27:    H: 18 L: 4 : 1%  | Partly cloudy and a gentle breeze\nMonday, 2022-03-28:    H: 18 L: 9 : 5%  | Light cloud and light winds\nTuesday, 2022-03-29: 殺  H: 11 L: 6 : 64% | Light rain and a gentle breeze\n<small>Last Updated: 17:02:52.807</small>"}
```

To ease setting up location configs, there is also `bbc-weather-cli location --search=`, which returns the BBC location code in the first textual column.

``` shell
$ bbc-weather-cli loc -s london
Total Results:  15
2643743 London, Greater London  Type: settlement
6058560 London, Canada  Type: settlement
2643736 Londonderry, Derry City And Strabane    Type: settlement
...

```

There's also a basic CLI output mode, which prints a basic summary of the current weather report to stdout.

``` shell
$ bbc-weather-cli cli
London, Greater London:   11  11 kph E : 1% | Sunny intervals and light winds
```

To use a custom configuration location, you can add `--conf=path/to/config.json` to the default, json, or cli arg modes.

``` shell
$ bbc-weather-cli json --conf=test-config.json
{"text":"  18","alt":" 10 kph N : 9% | Light cloud and light winds","tooltip":"<big>Melbourne, Australia</big>\nNow:  10 kph N : 9% | Light cloud and light winds\n\nSunday, 2022-03-27:    H: 25 L: 18 : 10% | Light cloud and a gentle breeze\nMonday, 2022-03-28:    H: 25 L: 15 : 15% | Light cloud and a moderate breeze\nTuesday, 2022-03-29:   H: 22 L: 14 : 10% | Light cloud and a moderate breeze\n<small>Last Updated: 03:16:44.396</small>"}

$ bbc-weather-cli cli -ctest.json
Melbourne, Australia:   18  10 kph N : 9% | Light cloud and light winds
```

## Configuration

`bbc-weather-cli` expects a `config.json` file that holds the JSON API web endpoint location, and the weather settlement's BBC location code. It will search for a `config.json` file at the following locations in order:

1. `$XDG_CONFIG_HOME/bbc-weather-cli/config.json` -- (Usually `$HOME/.config/bbc-weather-cli/config.json`)
2. The user data directory -- (Usually `$HOME/.bbc-weather-cli/config.json`)
3. The home directory -- (Usually `$HOME/.bbc-weather-cli-config.json`)

These paths will differ on other OSes (e.g. Windows or MacOS).
If multiple config files exist, the first location will be prioritised. I.e., a config file in `XDG_CONFIG_HOME` will be used even if one exists in user data.

If no config files exist, the program will create one on first run in `XDG_CONFIG_HOME`. This default config will not have a location code - please set the `forecastLocationCode` JSON parameter with a string value of the BBC location code. E.g.:

``` json
{
    "locationEndpoint": ... ,
    ...
    "forecastLocationCode": "2158177" // Set this value 
}
```

## Building

Although Haskell is pure joy to work with, Haskell's build tools are not quite so joyous. Instructions follow for all the Haskell build tools supported.

If you're building on Windows, stack is probably your best bet.

### Nix 

Simple build:

``` shell
$ nix-build release.nix
...
$ ./result/bin/bbc-weather-cli 
```

To install it for your user/system on Nix (single-user) or NixOS, I recommend adding this package to a custom overlay and using `nix-env`, `configuration.nix`, or `home-manager`.

### Cabal

``` shell
$ cabal build ## Builds into the dist-newstyle folder.
              ## Copy the binary into $HOME/.local/bin if desired
```

### Stack

``` shell
$ stack build
$ stack install ## Installs into $HOME/.local/bin by default
```

### Stack + Nix

``` shell
$ stack --nix build
$ stack install
```

Depending on your setup, you may need to enter a `nix-shell` with `pkgs.zlib` as a build input.

### Haskell Platform (NOTE: Deprecated)

``` shell
$ cd app
$ ghc Main.hs -o bbc-weather-cli
```

## Changelog

Please see CHANGELOG.md


CHANGELOG.md
# Revision history for bbc-weather-cli

## 0.5.0.0 -- 2022-03-27

* Added basic cli mode

## 0.4.0.0 -- 2022-03-22

* Added default parameter for number of summary days to display, `defaultSummaryDays`
* Added cmdArgs for a CLI interface
* Fixed JSON parsing bug where `pollenIndexBand` and `pollutionIndexBand` were being parsed as integers
* Added cli location query options with `bbc-weather-cli loc --search=`
* Added support for custom configuration file locations `bbc-weather-cli json --conf=` and `bbc-weather-cli config --conf=`


app/Main.hs
module Main where

import JSONTypes
import Utils
import qualified MainArgs as ARG
import System.Console.CmdArgs (cmdArgsRun)

import Network.HTTP.Client
import Network.HTTP.Client.TLS
import Network.HTTP.Types.Status (statusCode)
import System.Directory (doesFileExist, canonicalizePath)
import Control.Monad.Trans.Except
import Control.Exception (catch, IOException, Exception, SomeException)
import Control.Monad.Catch (catchAll)
import System.Exit (die)
import Data.Aeson (decode, encode, eitherDecode)
import qualified Data.Time.Clock as Time
import qualified Data.Time.LocalTime as Local
import qualified Data.ByteString.Lazy as BL
import qualified Data.ByteString.Lazy.Char8 as BL8
import qualified Data.ByteString as BS
import qualified Data.Time.Calendar as Cal (Day, diffDays, dayOfWeek)
import Data.Aeson.Encode.Pretty (encodePretty)
import Data.List (sortOn, intersperse)
import Data.List.Split (splitOn)
import Data.Maybe (fromJust)
import qualified Network.URI.Encode as URIEnc (encode)

configFile :: IO String
configFile = do
  existingConf <- getExistingConfFilePath
  if existingConf == ""
    then do
          fp <- makeDefaultPreferredConfFilePath
          BL8.writeFile fp $ encodePretty defaultConfig
          return fp
    else return existingConf


readConfigFile :: String -> ExceptT IOException IO BL.ByteString
readConfigFile f =  ExceptT ((Right <$> BL.readFile f) `catch` (\e -> return . Left $ e))

getConfig :: IO ConfigFile
getConfig = configFile >>= getCustomConfig

getCustomConfig :: String -> IO ConfigFile
getCustomConfig confLoc = do
    canConfLoc <- canonicalizePath confLoc
    confFileExists <- doesFileExist canConfLoc
    if not confFileExists
      then (die $ "Config file at: " ++ canConfLoc ++ " does not exist. Unable to continue.")
      else runExceptT (readConfigFile canConfLoc) >>=
            either printErrorAndExit (return . eitherDecode) >>=
              either printDecodeErrorAndExit return

addQueryParam :: BS.ByteString -> String -> String -> BS.ByteString
addQueryParam query param value = BS.append query $
                                    BL8.toStrict $
                                      BL8.pack $
                                        "&" ++ param ++ "=" ++ value

queryForBBCLocation :: String -> IO (Maybe BL.ByteString)
queryForBBCLocation loc = do
  conf <- getConfig
  man <- newTlsManager
  req <- (parseRequest $ "GET " ++ (locationEndpoint conf)) `catchAll` printUrlParsingErrorAndExit
  let req1 = req {
              queryString = addQueryParam (addQueryParam (queryString req) "api_key" (locationApiKey conf))
                                          "s" (URIEnc.encode loc)
              }
   in do
    res <- httpLbs req1 man
    case statusCode $ responseStatus res of
      200 -> return $ Just $ responseBody res
      _   -> return Nothing

bbcLocationToString :: BBCLocation -> String
bbcLocationToString (BBCLocation iD name container _ _ _ _ _ _ typ) = iD ++ "\t" ++ name ++ ", " ++ container ++ "\tType: " ++ typ

parseBBCLocResults :: Maybe BL.ByteString -> Maybe (Integer, [String])
parseBBCLocResults Nothing = Nothing
parseBBCLocResults (Just res) = let res1 = decode res :: Maybe BBCLocResponse
                                 in case res1 of
                                      Nothing -> Nothing
                                      Just res2 -> Just $
                                                    ((bbclq_totalResults . bbclres_results . bbclr_response $ res2),
                                                     bbcLocationToString <$> (bbclq_results . bbclres_results . bbclr_response $ res2) )

displayBBCLocResults :: Maybe (Integer, [String]) -> IO ()
displayBBCLocResults Nothing = putStrLn "Location search failure"
displayBBCLocResults (Just (total, res)) = putStrLn ("Total Results:\t" ++ (show total)) >>
                                            mapM_ putStrLn res

queryLocationAndDisplay :: String -> IO ()
queryLocationAndDisplay = \x -> queryForBBCLocation x >>= displayBBCLocResults . parseBBCLocResults

printErrorAndExit :: Exception e => e -> IO a
printErrorAndExit e = putStrLn "Encountered error:" >> die (show e)

printDecodeErrorAndExit :: String -> IO a
printDecodeErrorAndExit e = putStrLn "Unable to continue: error when decoding config.json" >> die e

printUrlParsingErrorAndExit :: SomeException -> IO a
printUrlParsingErrorAndExit e = putStrLn "Error when parsing the URL from config.json" >> die (show e)

fetchWeatherForecast :: Manager -> String -> String -> IO (Maybe BL.ByteString)
fetchWeatherForecast man endp loc = do
  req <- (parseRequest $ "GET " ++ endp ++ loc) `catchAll` printUrlParsingErrorAndExit
  res <- httpLbs req man
  case statusCode $ responseStatus res of
    200 -> return $ Just $ responseBody res
    _   -> return Nothing

fetchWeatherForecastFromConfig :: Manager -> ConfigFile -> IO (Maybe BL.ByteString)
fetchWeatherForecastFromConfig man conf = fetchWeatherForecast man (forecastEndpoint conf) (forecastLocationCode conf)

fetchAggregateForecastFromConfig :: Manager -> ConfigFile -> IO (Maybe BL.ByteString)
fetchAggregateForecastFromConfig man conf = fetchWeatherForecast man (aggregateForecastEndpoint conf) (forecastLocationCode conf)

getWeatherObservations :: Weather -> [WeatherData]
getWeatherObservations w = let feats = filter (\x -> (fp_observations . f_properties) x /= Nothing) $ w_features w
                            in case length feats > 0 of
                                 True -> fromJust $ fp_observations . f_properties $ head feats
                                 False -> []

getWeatherForecasts :: Weather -> [WeatherData]
getWeatherForecasts w = let feats = filter (\x -> (fp_forecasts . f_properties) x /= Nothing) $ w_features w
                         in case length feats > 0 of
                              True -> fromJust $ fp_forecasts . f_properties $ head feats
                              False -> []

getClosestWeatherData :: Time.UTCTime -> [WeatherData] -> WeatherData
getClosestWeatherData now w = head $ sortOn (\x -> abs . Time.nominalDiffTimeToSeconds . ((flip Time.diffUTCTime) now) . time_utc . wdata_time $ x) w

convertDetailedTimeStringToUTC :: Local.TimeZone -> DetailedReport -> Time.UTCTime
convertDetailedTimeStringToUTC tz dr = Local.zonedTimeToUTC $
                                         Local.ZonedTime (Local.LocalTime (dr_localDate dr) (Local.timeToTimeOfDay $
                                                                                               Time.secondsToDiffTime $
                                                                                                 60 * 60 * (read (take 2 $ dr_timeslot dr)))) tz

getLocalDay :: IO Cal.Day
getLocalDay = Local.getZonedTime >>=
                 return . Local.zonedTimeToLocalTime >>=
                   return . Local.localDay

getClosestDetailedData :: Local.TimeZone -> Time.UTCTime -> AggregateWeather -> DetailedReport
getClosestDetailedData tz now w = let alldr = concat $ df_reports . awf_detailed <$> forecasts w
                                   in head $
                                        sortOn (\x -> abs . Time.nominalDiffTimeToSeconds .
                                                  ((flip Time.diffUTCTime) now) . convertDetailedTimeStringToUTC tz $ x) alldr

getNextNSummaryData :: Cal.Day -> Int -> AggregateWeather -> [SummaryReport]
getNextNSummaryData today n w = let summs = (sf_report . awf_summary) <$> forecasts w
                                    day_diff_amount = ((flip Cal.diffDays) today) . sr_localDate
                                 in take n $
                                      sortOn day_diff_amount $
                                        filter (\x -> day_diff_amount x >= 0) summs

aggregateWeatherToLocationString :: AggregateWeather -> String
aggregateWeatherToLocationString = concat . (intersperse ", ") . ((<*>) [(awl_name . location), (awl_container . location)]) . pure

aggregateWeatherToLastUpdatedTimeOfDay :: AggregateWeather -> String
aggregateWeatherToLastUpdatedTimeOfDay = show . Local.localTimeOfDay . Local.zonedTimeToLocalTime . lastUpdated


summaryReportToTooltipLine :: SummaryReport -> String
summaryReportToTooltipLine sr = (show . Cal.dayOfWeek $ sr_localDate sr) ++
                                  ", " ++
                                    (show $ sr_localDate sr) ++
                                      ": " ++
                                        (getWeatherTypeIcon $ sr_weatherType sr) ++ "  " ++
                                          "H: " ++ (show $ sr_maxTempC sr) ++ "\xe339 " ++
                                          "L: " ++ (show $ sr_minTempC sr) ++ "\xe339 " ++
                                          "\xe371: " ++ (show $ sr_precipitationProbabilityInPercent sr) ++ "%" ++
                                            " | " ++ (sr_enhancedWeatherDescription sr)

equalizeStringChunksBySep :: String -> [String] -> [String]
equalizeStringChunksBySep "" ttls = ttls
equalizeStringChunksBySep sep ttls = let maxlen = maximum $ length . head <$> (splitOn sep) <$> ttls
                                         padhead x = [head x ++ sep ++ (replicate  (maxlen - (length $ head x))  ' ')] <> (intersperse sep $ tail x)
                                      in concat . padhead <$> (splitOn sep <$> ttls)

summaryReportsToTooltipSummary :: [SummaryReport] -> String
summaryReportsToTooltipSummary = concat . intersperse "\n" .
                                  equalizeStringChunksBySep "%" .
                                   equalizeStringChunksBySep ":" .
                                    (<$>) summaryReportToTooltipLine

aggregateWeatherToTooltip :: Cal.Day -> String -> AggregateWeather -> String
aggregateWeatherToTooltip today dr_str aw = "<big>" ++ (aggregateWeatherToLocationString aw) ++ "</big>\n" ++
                                              "Now: " ++ dr_str ++ "\n\n" ++
                                                (summaryReportsToTooltipSummary $ getNextNSummaryData today 3 aw) ++ "\n" ++
                                                  "<small>Last Updated: " ++ (aggregateWeatherToLastUpdatedTimeOfDay aw) ++ "</small>"

aggregateNWeatherToTooltip :: Integral n => n -> Cal.Day -> String -> AggregateWeather -> String
aggregateNWeatherToTooltip n today dr_str aw = "<big>" ++ (aggregateWeatherToLocationString aw) ++ "</big>\n" ++
                                                "Now: " ++ dr_str ++ "\n\n" ++
                                                   (summaryReportsToTooltipSummary $ getNextNSummaryData today (fromIntegral n) aw) ++ "\n" ++
                                                     "<small>Last Updated: " ++ (aggregateWeatherToLastUpdatedTimeOfDay aw) ++ "</small>"

getSwitchConf :: String -> IO ConfigFile
getSwitchConf "DEFAULT" = getConfig
getSwitchConf x = getCustomConfig x

getSwitchConfFilePath :: String -> IO String
getSwitchConfFilePath "DEFAULT" = getExistingConfFilePath
getSwitchConfFilePath x = canonicalizePath x

mainGetAggregateWeather :: String -> IO AggregateWeather
mainGetAggregateWeather uConfLoc = do
  man <- newTlsManager
  conf <- getSwitchConf uConfLoc
  case forecastLocationCode conf of
    "" -> do
      confloc <- getSwitchConfFilePath uConfLoc
      die $ "No location code has been set in the config file at:\n\t" ++ confloc
    _  -> do
      w <- fetchAggregateForecastFromConfig man conf
      case w of
        Nothing -> die "BBC's weather API did not return a valid HTTP response. Unable to continue."
        Just w1 -> let w2 = decode w1 :: Maybe AggregateWeather
                    in case w2 of
                         Nothing -> die "BBC's weather API did not return a valid JSON payload. Unable to parse."
                         Just w3 -> return w3

waybar :: String -> IO ()
waybar uConfLoc = do
  agg <- mainGetAggregateWeather uConfLoc
  conf <- getSwitchConf uConfLoc
  tz  <- Local.getCurrentTimeZone
  now <- Time.getCurrentTime
  today <- getLocalDay
  let closestDetailed = getClosestDetailedData tz now agg
      summaryDays = case (defaultSummaryDays conf) of
                      Nothing -> defaultSummaryDaysValue
                      Just n  -> n
      jsonout = WaybarJSON (detailedReportToMainStatus closestDetailed)
                           (detailedReportToAltStatus closestDetailed)
                           (aggregateNWeatherToTooltip summaryDays today (detailedReportToAltStatus closestDetailed) agg)
   in BL8.putStr . encode $ jsonout

cliWeatherStatus :: String -> IO ()
cliWeatherStatus uConfLoc = do
  agg <- mainGetAggregateWeather uConfLoc
  conf <- getSwitchConf uConfLoc
  tz  <- Local.getCurrentTimeZone
  now <- Time.getCurrentTime
  today <- getLocalDay
  putStr $ aggregateWeatherToLocationString agg ++ ": "
  putStrLn $ detailedReportToCliStatus $ getClosestDetailedData tz now agg

displayConfig :: IO ()
displayConfig = do
  discoveredConf <- getExistingConfFilePath
  case discoveredConf of
    "" -> die "No existing config.json files were found!\nA default config.json file will be created on first run"
    x -> do
          putStrLn $ "Found an existing config file at: " ++ x ++ "\n"
          conf <- getConfig
          BL8.putStrLn $ encodePretty conf

displayCustomConfig :: String -> IO ()
displayCustomConfig "DEFAULT" = displayConfig
displayCustomConfig confLoc = getCustomConfig confLoc >>=
                               \conf -> BL8.putStrLn $ encodePretty conf

checkConfigLocation :: ConfigFile -> IO Bool
checkConfigLocation (ConfigFile _ _ _ locCode agg _) = do
  putStrLn "Assert: location code is not blank"
  case locCode of
    "" -> putStrLn "FAIL: No location code has been declared.\n \
                     \ Please set a location code in config.json\n \
                     \ Run bbc-weather-cli loc --search= if you don't know your location code" >>
            return False
    lC -> do
      putStrLn $ "OK: Location code is " ++ lC
      man <- newTlsManager
      res <- fetchWeatherForecast man agg lC
      case res of
        Nothing -> do
          putStrLn $ "FAIL: The endpoint " ++ agg ++ lC ++ " did not return an HTTP 200 response."
          return False
        Just _ -> do
          putStrLn "OK: The aggregate weather forecast returned an HTTP 200 response"
          putStrLn "PASSED: Location Code check succeeded!"
          return True


checkConfig :: ConfigFile -> IO ()
checkConfig conf = checkConfigLocation conf >>= print

handleArgs :: ARG.BBCCli -> IO ()
handleArgs (ARG.Json "waybar" conf) = waybar conf
handleArgs (ARG.Json _ _) = die "Other JSON formats are not supported yet"
handleArgs (ARG.Config confLoc False) = displayCustomConfig confLoc
handleArgs (ARG.Config "DEFAULT" True) = getConfig >>= checkConfig
handleArgs (ARG.Config confLoc True) = getCustomConfig confLoc >>= checkConfig
handleArgs (ARG.Location "") = die "No search parameter provided for the location query"
handleArgs (ARG.Location loc) = queryLocationAndDisplay loc
handleArgs (ARG.Cli confLoc) = cliWeatherStatus confLoc
handleArgs _ = putStrLn "No supported arguments given"

main :: IO ()
main = cmdArgsRun ARG.bbccli >>= handleArgs


app/MainArgs.hs
{-# LANGUAGE DeriveDataTypeable #-}

module MainArgs where
import System.Console.CmdArgs

version :: String
version = "0.5.0.0"

data BBCCli = Json { format :: String
                   , conf :: String
                   }
            | Config { conf :: String
                     , check :: Bool
                     }
            | Location { search :: String }
            | Cli { conf :: String }
            deriving (Data, Typeable, Show, Eq)

json = Json { format = "waybar" &= help "Specify a specific output format (e.g. waybar)\
                                         \ \nDefaults to format=waybar" &= opt "waybar"
            , conf = confFlag "DEFAULT"
            } &= help "Fetches and prints the weather to stdout" &= auto

confFlag x = x &= help "Path to a custom config file" &= opt "DEFAULT" &= typFile

config = Config { conf = confFlag "DEFAULT"
                , check = def &= help "Check config.json for potential issues"
                } &= help "View and modify bbc-weather-cli's config"

location = Location { search = def &= help "The string to search by\nE.g. --search=\"london\""
                    } &= help "Query the BBC Location API with a search string"

cli = Cli { conf = confFlag "DEFAULT" }

bbccli = cmdArgsMode $ modes [json, config, location, cli]
  &= helpArg [help "Fetch and display the weather, or modify configurations"
             , name "h"]
  &= program "bbc-weather-cli"
  &= summary ("bbc-weather-cli v." ++ version ++ "\nFetches and displays the weather from BBC's Weather API")

runBbccli = cmdArgsRun bbccli >>= print


app/Utils.hs
module Utils (getXdgConfDir,
              getXdgConfFile,
              getUserdataConfDir,
              getUserdataConfFile,
              getHomeConfFile,
              getExistingConfFilePath,
              makeDefaultPreferredConfFilePath,
              getWeatherTypeIcon,
              detailedReportToMainStatus,
              detailedReportToAltStatus,
              detailedReportToCliStatus) where

import JSONTypes

import System.Directory
import System.IO (FilePath)
import Control.Monad (filterM)
import Data.List (intersperse)

bbcConfDir :: String
bbcConfDir = "bbc-weather-cli"

bbcConfFile :: String
bbcConfFile = "config.json"

getXdgConfDir :: IO FilePath
getXdgConfDir = getXdgDirectory XdgConfig bbcConfDir

getXdgConfFile :: IO FilePath
getXdgConfFile = getXdgConfDir >>= return . (\x -> x ++ "/" ++ bbcConfFile)

getUserdataConfDir :: IO FilePath
getUserdataConfDir = getAppUserDataDirectory bbcConfDir

getUserdataConfFile :: IO FilePath
getUserdataConfFile = getUserdataConfDir >>= return . (\x -> x ++ "/" ++ bbcConfFile)

getHomeConfFile :: IO FilePath
getHomeConfFile = getHomeDirectory >>= return . (\x -> x ++ "/." ++ bbcConfDir ++ "-" ++ bbcConfFile)

ioConfFiles :: IO [FilePath]
ioConfFiles = sequence [getXdgConfFile, getUserdataConfFile, getHomeConfFile]

anyConfFilesExist :: IO Bool
anyConfFilesExist = ioConfFiles >>=
                      mapM doesFileExist >>=
                        return . any id

getExistingConfFilePath :: IO FilePath
getExistingConfFilePath = ioConfFiles >>= \iof -> anyConfFilesExist >>=
                            \exists -> if exists
                                        then filterM doesFileExist iof >>= return . head
                                        else return ""

makeDefaultPreferredConfFilePath :: IO FilePath
makeDefaultPreferredConfFilePath = getXdgConfDir >>=
                                     createDirectoryIfMissing True >>
                                       getXdgConfFile

getWeatherTypeIcon :: Integer -> String
getWeatherTypeIcon 0  = "\xf186"   --  - Clear sky (night)
getWeatherTypeIcon 1  = "\xe30d"   --  - Sunny day
getWeatherTypeIcon 2  = "\xe37e"   --   - Partly cloudy (night)
getWeatherTypeIcon 3  = "\xe30c"   --  - Sunny intervals
getWeatherTypeIcon 4  = "\xe37a"   --   - Sandstorm
getWeatherTypeIcon 5  = "\xfa90"   -- 敖 - Mist day / night
getWeatherTypeIcon 6  = "\xfa90"   -- 敖 - Fog day / night
getWeatherTypeIcon 7  = "\xe376"   --  - White medium level cloud
getWeatherTypeIcon 8  = "\xe312"   --  - Black low level cloud
getWeatherTypeIcon 9  = "\xfa96"   -- 殺 - Light rain shower (night)
getWeatherTypeIcon 10 = "\xe30b"   --  - Light rain shower (day)
getWeatherTypeIcon 11 = "\xe34a"   --  -Drizzle
getWeatherTypeIcon 12 = "\xfa96"   -- 殺 - Light rain day or night
getWeatherTypeIcon 13 = "\xe325"   --  - Heavy rain shower (night)
getWeatherTypeIcon 14 = "\xe308"   --  - Heavy rain shower (day)
getWeatherTypeIcon 15 = "\xfa95"   -- 歹 -  Heavy rain
getWeatherTypeIcon 16 = "\xe3ac"   --  - Sleet shower (night)
getWeatherTypeIcon 17 = "\xe3aa"   --  - Sleet shower (day)
getWeatherTypeIcon 18 = "\xe3ad"   --  - Cloudy with sleet
getWeatherTypeIcon 19 = "\xe321"   --  - Hail shower (night)
getWeatherTypeIcon 20 = "\xe304"   --  - Hail shower (day)
getWeatherTypeIcon 21 = "\xe314"   --  - Cloudy with hail
getWeatherTypeIcon 22 = "\xe327"   --  - Light snow shower (night)
getWeatherTypeIcon 23 = "\xe30a"   --  - Light snow shower (day)
getWeatherTypeIcon 24 = "\xe31a"   --  - Cloudy with light snow
getWeatherTypeIcon 25 = "\xe367"   --  - Heavy snow shower (night)
getWeatherTypeIcon 26 = "\xe365"   --  - Heavy snow shower (day)
getWeatherTypeIcon 27 = "\xfa97"   -- 流 - Cloudy with heavy snow
getWeatherTypeIcon 28 = "\xe337"   --  - Thundery shower (night)
getWeatherTypeIcon 29 = "\xe30e"   --  - Thundery shower (day)
getWeatherTypeIcon 30 = "\xe31d"   --  - Thunderstorms
getWeatherTypeIcon 31 = "\xe208"   --  - Tropical storm (bbc uses some type of spinning thing)
getWeatherTypeIcon 32 = "HAZY"     --  - Hazy
getWeatherTypeIcon 33 = "SANDSTORM"-- Text - Sandstorm
getWeatherTypeIcon 34 = "MIST"     -- Text - Mist
getWeatherTypeIcon 35 = "FOG"      -- Text - Fog
getWeatherTypeIcon 36 = "\xe33d"   --  - Light cloud
getWeatherTypeIcon 37 = "\xf0c2"   --  - Thick cloud
getWeatherTypeIcon 38 = "\xe239"   --  - Thick cloud and drizzle
getWeatherTypeIcon 39 = "\xfa96"   -- 殺 - Thick cloud and raindrop
getWeatherTypeIcon 40 = "\xfa96\xe34a"  -- 殺 - Thick cloud and double raindrop
getWeatherTypeIcon 41 = "\xfb7d"   -- ﭽ - Thick cloud, snowy + rainy
getWeatherTypeIcon 42 = "\xfa91"   -- 晴 - Thick cloud and hailstones
getWeatherTypeIcon 43 = "\xfa97"   -- 流 - Thick cloud and snowflake
getWeatherTypeIcon 44 = "\xe31a"   --  - Thick cloud and double snowflake
getWeatherTypeIcon 45 = "\xfa92"   -- 朗 - Thick cloud and lightning
getWeatherTypeIcon 46 = "\xe208"   --  - Tropical storm again
getWeatherTypeIcon 47 = "HAZY"    -- Text - Hazy
getWeatherTypeIcon _  = "\xf420"   --  - Unknown status

detailedReportToMainStatus :: DetailedReport -> String
detailedReportToMainStatus dr = (getWeatherTypeIcon $ dr_extendedWeatherType dr) ++
                                  "  " ++
                                    (show $ dr_temperatureC dr) ++
                                      "\xe339" -- 

detailedReportToAltStatus :: DetailedReport -> String
detailedReportToAltStatus dr = "\xe3a9" ++
                                 " " ++ (show $ dr_windSpeedKph dr) ++ " kph " ++
                                   (dr_windDirectionAbbreviation dr) ++ " " ++
                                     "\xe371:" ++ " " ++ (show $ dr_precipitationProbabilityInPercent dr) ++
                                       "% | " ++ (dr_enhancedWeatherDescription dr)

detailedReportToCliStatus :: DetailedReport -> String
detailedReportToCliStatus = concat . (intersperse " ") . ((<*>) [detailedReportToMainStatus, detailedReportToAltStatus]) . pure
